tools/crm_shadow.c

```diff
@@ -93,453 +145,574 @@ shadow_teardown(char *name)
-        pcmk__option_paragraph
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        " crm_shadow --delete myShadow --force", pcmk__option_example
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "Upload current shadow configuration (named myShadow) "
-            "to running cluster:",
-        pcmk__option_paragraph
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        " crm_shadow --commit myShadow", pcmk__option_example
-    },
-    { 0, 0, 0, 0 }
+static bool
+cmd_is_dangerous(enum shadow_command cmd)
+{
+    switch (cmd) {
+        case shadow_cmd_commit:
+        case shadow_cmd_delete:
+            return true;
+        default:
+            return false;
+    }
+}
+
+static gboolean
+which_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+         GError **error)
+{
+    SET_COMMAND(shadow_cmd_which);
+    return TRUE;
+}
+
+static gboolean
+display_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+           GError **error)
+{
+    SET_COMMAND(shadow_cmd_display);
+    return TRUE;
+}
+
+static gboolean
+diff_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+        GError **error)
+{
+    SET_COMMAND(shadow_cmd_diff);
+    return TRUE;
+}
+
+static gboolean
+file_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+        GError **error)
+{
+    SET_COMMAND(shadow_cmd_file);
+    return TRUE;
+}
+
+static gboolean
+create_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+          GError **error)
+{
+    SET_COMMAND(shadow_cmd_create);
+    pcmk__str_update(&options.shadow, optarg);
+    return TRUE;
+}
+
+static gboolean
+create_empty_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+                GError **error)
+{
+    SET_COMMAND(shadow_cmd_create_empty);
+    pcmk__str_update(&options.shadow, optarg);
+    return TRUE;
+}
+
+static gboolean
+commit_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+          GError **error)
+{
+    SET_COMMAND(shadow_cmd_commit);
+    pcmk__str_update(&options.shadow, optarg);
+    return TRUE;
+}
+
+static gboolean
+delete_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+          GError **error)
+{
+    SET_COMMAND(shadow_cmd_delete);
+    pcmk__str_update(&options.shadow, optarg);
+    return TRUE;
+}
+
+static gboolean
+edit_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+        GError **error)
+{
+    SET_COMMAND(shadow_cmd_edit);
+    return TRUE;
+}
+
+static gboolean
+reset_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+         GError **error)
+{
+    SET_COMMAND(shadow_cmd_reset);
+    pcmk__str_update(&options.shadow, optarg);
+    return TRUE;
+}
+
+static gboolean
+switch_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+          GError **error)
+{
+    SET_COMMAND(shadow_cmd_switch);
+    pcmk__str_update(&options.shadow, optarg);
+    return TRUE;
+}
```

Comment:
Yeah that's a good thought and I was on the fence about that.

On the one hand, the code is shorter.

On the other, it introduces a lot of redundant checks -- the option parser already tells us what we need to do, basically, and we can jump into a callback written for that instead of checking against every string again until we find a match.

Back to the first hand, we're not really worried about performance during option parsing.

Reply:
@kgaillot any preference here, I don't really care

Reply:
Either is fine

Reply:
gonna do the strcmp